cmdx.py

```diff
@@ -3861,13 +3948,144 @@ def _python_to_mod(value, plug, mod):
+        else:
+            sup = Component
+
+        self = super(MetaComponent, sup).__call__(mdagpath, mobject, exists, modifier)
+        return self
+
+
+@add_metaclass(MetaComponent)
+class Component(object):
+    _Fn = om.MFnComponent
+
+    typeName = None
+
+    def __init__(self, node, mobject, exists=True, modifier=None):
+        """A Maya component container
+
+        Arguments:
+            node (DagNode): cmdx shape instance
+            mobject (maya.api.OpenMaya.MObject): Internal component object
+        """
+        assert isinstance(node, DagNode), "%s is not a DagNode" % node
+        assert isinstance(mobject, om.MObject), "%s is not a MObject" % mobject
+
+        self._node = node
+        self._mdagpath = node.dagPath()
+
+        self._mobject = mobject
+        self._fn = self._Fn(mobject)
+
+        self._it = None
+
+    def __str__(self):
+        """Return component selection string
+        """
+        sl = om.MSelectionList()
+        sl.add((self._mdagpath, self._mobject))
+        return ' '.join(sl.getSelectionStrings(0))
+
+    path = __str__
+
+    def __nonzero__(self):
+        return not self._fn.isEmpty
+
+    __bool__ = __nonzero__
+
+    @property
+    def complete(self):
+        return self._fn.isComplete
+
+
+class Component1D(Component):
+    _Fn = om.MFnSingleIndexedComponent
+
+    def __len__(self):
+        return self._fn.getCompleteData()
+
+    def __contains__(self, item):
+        return item in self._fn.getElements()
+
+    def __iter__(self):
+        """Iterate over the component indices
+        """
+        if self._fn.isComplete:
+            for i in xrange(len(self)):
+                yield i
+        else:
+            for i in self._fn.getElements():
+                yield i
+
+    def __getitem__(self, index):
+        cps = self._Fn().create(self._type)
+        fn = self._Fn(cps)
+        n = self.maxLength()
+        if n is None:
+            raise IndexError("{} doesn't support get item".format(self.__class__))
+
+        if isinstance(index, slice):
+            fn.addElements(range(*index.indices(n)))
+        elif isinstance(index, tuple):
+            fn.addElements(index)
+        else:
+            if index < 0:
+                index = n + 1 + index
+            fn.addElement(index)
+
+        return self.__class__(self._node, cps)
+
+    def iterate(self):
+        """Iterate over the component iterator from the indices list"""
+        it = self._node.iterator()
+
+        for i in self:
+            it.setIndex(i)
+            yield i
+
+    def append(self, item):
+        if not isinstance(item, int):
+            raise TypeError()
+        if item >= self.maxLength():
+            raise IndexError()
+        if not self._fn.isComplete and item not in self:
+            self._fn.addElement(item)
+
+    @property
+    def elements(self):
+        return self._fn.getElements()
+
+
+class MeshVertex(Component1D):
+    _type = om.MFn.kMeshVertComponent
+
```

Comment:
Can we make `MeshVertex` and `Component1D` private? Do they need to be accessed independently? Let's call these `_MeshVertex` and `_Component1D` to let the user know these are internal-only.

Reply:
yep you're right! there is no point to cast Component1D and MeshVertex directly

Reply:
I made only _Component1D private. It should be possible to test instance type of casted Component with `isistance(cmdx.MeshVertex, object)` (or we have to implement something like `Node.isA`)